crates/subspace-service/src/lib.rs

```diff
@@ -561,6 +572,11 @@ where
 
                 let forward = polkadot_overseer::forward_events(
                     client,
+                    Box::pin(
+                        imported_block_notification_stream
```

Comment:
I believe this will still be racy even if we initialize config in time as you mentioned in one of the commits messages. The reason is that this code may miss some messages in `imported_block_notification_stream` before outer future here is actually polled and subscription is created. So subscription needs to be created early on before networking of the primary node even starts or executor logic needs to be able to keep track of blocks it has produced and being able to process gaps (like farmer does for instance), so neither config initialization nor time of creation of this subscription would matter.

Reply:
>The reason is that this code may miss some messages in imported_block_notification_stream before outer future here is actually polled and subscription is created.

Is there anything I can experiment with to confirm this? I'm not sure that messages will be missed if you didn't subscribe it earlier. Based on my debugging of the problem, no messages will be missed, some early messages are just ignored because of this line that we only handle the message when the config has been initialized/is`Some`. https://github.com/subspace/subspace/blob/4929d4d99ae4be7fa0bc7eef13852a48e9bae1a7/polkadot/node/overseer/src/lib.rs#L359

Reply:
I'm thinking of `start_network.start_network()` call and the fact that it is earlier in the list of things that are executed. So there is a hypothetical possibility for it to connect to some node and import one block before the subscription to block import is created. Just imagine arbitrary delay at any place in the code and thing it it'll still work properly when delay is introduced. I think it will not in this case, thus must be racy.

Reply:
It comes to my mind that invoking `start_network()` of the primary node at the end of everything should also solve your concern?

https://github.com/subspace/subspace/blob/4929d4d99ae4be7fa0bc7eef13852a48e9bae1a7/crates/subspace-service/src/lib.rs#L504

This could be a simpler one.

Reply:
Yes, that is what I'll do in upcoming refactoring